File: polyglot/DesignPatterns/Singleton/singleton.cpp
#include <iostream>
#include <memory>

class SingleObject;
typedef std::shared_ptr<SingleObject> pSingleObject;

class SingleObject {
  private:
    SingleObject(){}
    static pSingleObject mInstance;
  public:
    static pSingleObject getInstance(){
      if(mInstance == nullptr){
        mInstance = pSingleObject(new SingleObject());
      }
      return mInstance;
    }
    void showMessage() const{
      std::cout << "The single object is alive and well\n";
    }
};

// definition of static member
pSingleObject SingleObject::mInstance = nullptr;


int main(int argc, char* argv[]){
  // will not compile
  // SingleObject nope;

  pSingleObject object1 = SingleObject::getInstance();
  object1->showMessage();

  pSingleObject object2 = SingleObject::getInstance();
  if(object1 == object2){ // comparison between smart pointers
    std::cout << "The two objects are the same\n";
  }

  return 0;
}

File: polyglot/DesignPatterns/Singleton/singleton.clj
; Singleton design pattern
(ns singleton (:gen-class))

(def SingleObject   ; 
  (letfn
    ; instance interface
    [(instance[m]
        (cond (= m :showMessage) (showMessage)
               :else (println "SingleObject: unknown instance operation error")))
     (showMessage []
       (println "The single object is alive and well"))
    ; static interface
     (static [m]
       (cond (= :getInstance) (getInstance)
             :else (println "SingleObject: unknown static operation error")))
     (getInstance []
       instance)]
    ; returning static interface
    static))

(defn -main []
(def object1 (SingleObject :getInstance))
(object1 :showMessage)
(def object2 (SingleObject :getInstance))
(if (= object1 object2) (println "The two objects are the same")))

;(-main)






